ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java

```diff
@@ -70,26 +120,56 @@ public KafkaClientSupplier getKafkaClientSupplier() {
 
   @Override
   public SchemaRegistryClient getSchemaRegistryClient() {
-    return srClient;
+    return srClientSupplier.get();
   }
 
   @Override
   public Supplier<SchemaRegistryClient> getSchemaRegistryClientFactory() {
-    return srClientFactory;
+    return srClientSupplier;
   }
 
   @Override
   public ConnectClient getConnectClient() {
-    return connectClient;
+    return connectClientSupplier.get();
   }
 
   @Override
   public SimpleKsqlClient getKsqlClient() {
-    return ksqlClient;
+    return ksqlClientSupplier.get();
   }
 
   @Override
   public void close() {
-    adminClient.close();
+    if (((NotThreadSafeMemoizedSupplier)adminClientSupplier).isInitialized()) {
+      adminClientSupplier.get().close();
+    }
+  }
+
+  public static final class NotThreadSafeMemoizedSupplier<T> implements Supplier<T> {
```

Comment:
is it OK for this to be thread unsafe? If I understand correctly, and @spena can probably confirm, but I think we use `DefaultServiceContext` from multiple threads (we create one per request for the "user", but we also store one for the "server"). Either way, I think thread safety (using guava's memoized supplier) won't cause a noticeable performance overhead - a volatile field lookup in the average case is basically free - so we may as well do that unless we have justification otherwise

Reply:
+1 also.

Reply:
Good point. I think the reason @vpapavas created her own class was because we need to close the AdminClient.  If using Guava's version we'd have to `.get()` the instance to close it, which would create the instance if it hadn't already been created!  Basically, the issue is that there is no way to _know_ the instance had been created.  

I think a better solution would be to wrap the thread-safe guava memorizing supplier with something that tracks if its been initialized.  I'll knock up an example and add to my review.

Reply:
Yes, Andy's comment is correct. I talked with @spena about the thread safety issue and he said that there is only one thread per request. That's why I created the supplier as not thread safe.

Reply:
My concern was less that there's one thread per request, but rather that the ServiceContextFactory is shared across requests. @sergio do we create a new factory for every request (i.e. is `@Inject` called every request, or just `provide()`)? This is where I don't really have a working knowledge of the dependency injection framework that we use... either way, if there's no lock contention then synchronization is almost free and if there is thread contention then we need it 😂